x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -794,316 +768,298 @@ private void findTokenFromRefreshToken(String refreshToken, ActionListener<Searc
+     * return a user token and refresh token based on that. Otherwise this token document gets its refresh_token marked as refreshed, while
+     * also storing the Instant when it was refreshed along with a pointer to the new token document that holds the refresh_token that
+     * supersedes this one. The new document that contains the new access token and refresh token is created and finally the new access
+     * token and refresh token are returned to the listener.
      */
     private void innerRefresh(String tokenDocId, Map<String, Object> source, long seqNo, long primaryTerm, Authentication clientAuth,
-                              ActionListener<Tuple<UserToken, String>> listener, Iterator<TimeValue> backoff, Instant refreshRequested) {
+                              Iterator<TimeValue> backoff, Instant refreshRequested, ActionListener<Tuple<UserToken, String>> listener) {
         logger.debug("Attempting to refresh token [{}]", tokenDocId);
         final Consumer<Exception> onFailure = ex -> listener.onFailure(traceLog("refresh token", tokenDocId, ex));
-        final Optional<ElasticsearchSecurityException> invalidSource = checkTokenDocForRefresh(source, clientAuth);
-        if (invalidSource.isPresent()) {
-            onFailure.accept(invalidSource.get());
-        } else {
-            if (eligibleForMultiRefresh(source, refreshRequested)) {
-                final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
-                final String supersedingTokenDocId = (String) refreshTokenSrc.get("superseded_by");
-                logger.debug("Token document [{}] was recently refreshed, attempting to reuse [{}] for returning an " +
-                    "access token and refresh token", tokenDocId, supersedingTokenDocId);
-                final ActionListener<GetResponse> getSupersedingListener = new ActionListener<GetResponse>() {
-                    private final Consumer<Exception> maybeRetryOnFailure = ex -> {
-                        if (backoff.hasNext()) {
-                            final TimeValue backofTimeValue = backoff.next();
-                            logger.debug("retrying after [" + backofTimeValue + "] back off");
-                            final Runnable retryWithContextRunnable = client.threadPool().getThreadContext()
-                                    .preserveContext(() -> getTokenDocAsync(supersedingTokenDocId, this));
-                            client.threadPool().schedule(retryWithContextRunnable, backofTimeValue, GENERIC);
-                        } else {
-                            logger.warn("back off retries exhausted");
-                            onFailure.accept(ex);
-                        }
-                    };
+        final String refreshedNewTokenId;
+        try {
+            refreshedNewTokenId = checkTokenDocForRefresh(source, clientAuth);
```

Comment:
I much prefer the previous implementation with the optional. There is no need to throw and then catch to pass it on to a listener if we have a clean way of doing so.

Reply:
It's more than this going on behind the original implementation, I'll try to ~briefly~ explain.

Firstly, note that returning non-null, null, and throwing an exception are three different return results! The proposed change uses all of them.

The parsed token document is a map of maps. We don't parse it in a fully fledged entity.
The original implementation did some integrity checks (verify if entries existed and have the expected type) on the map(s) entries and then, subsequently, at completely different places in the code, values were extracted out of the maps, under the assumption that the null check had been performed before. I don't like that. I prefer to keep the null check as I navigate in the maps and extract the values. I feel like walking on a mine field when navigating maps knowing that null checks have been performed somewhere else. So, in short, the proposed implementation keeps value extracting paired with the null checks, without introducing a new entity.

Now, if we agree on this, the "value" extraction should be done once and grab everything in one pass, otherwise we'll have to duplicate the null checks on each value extraction, by the same philosophy. So, the proposed implementation grabs everything in one pass, and returns: `null` if the access token has not been refreshed before, non-null value representing the document id in case the access token has been refreshed recently, throw exception in case the token document is malformed or the refresh token is invalidated or expired. These are three return results, that the proposed implementation uses, again without introducing another class/entity.

What do you say? Can we keep it in the proposed form?

Reply:
> So, in short, the proposed implementation keeps value extracting paired with the null checks, without introducing a new entity.

What's the reasoning for not introducing a new entity? I've been thinking about this and from a code readability/understand-ability perspective a entity can convey the meaning of the returned result much better and enables someone to more easily dive into this code. In the past, I have definitely used this type of behavior but then I start to think about some of those times and we've actually gone back and changed it to have a result entity; examples would be AuthenticationResult and RoleRetrievalResult.

Reply:
> What's the reasoning for not introducing a new entity?

The "read the document once" was closer to the existing code, as the code at this abstraction level deals with maps, and only a single field is required from all the entity parsing anyway, so the parsing seemed like a stretch. Furthermore the field is not even an entity field, it's a document id, which should probably be abstracted away by the entity (as a token id).

Do you wish to go down the entity parsing path in this instance?

Reply:
I was thinking about something like:

```java
class RefreshStatus {

    boolean isRefreshed() { }

    boolean isRefreshable() { }

    boolean isValid() { }

    Exception getCause() { }

    String getRefreshedTokenId() { }
}
```

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/39808/commits/a7139a81a41e6c3ef1d76d3471ed6ae3dd5d63ec to assuage your concerns. The previous integrity checks would have returned 400 for actual 500 error codes. Given the previous feedback, I have changed them in this last "refactor".

@jaymode this is not exactly your suggestion because `isRefreshable` and `isValid` cannot cohabitate with `getCause` because `getCause` would be calling them internally so the caller of `getCause` would have no reason to call them on its own. Moreover, I believe `getCause()` is not how things are currently done in the code base, when we encounter validation errors we throw exceptions if the function is sync or call the `onFailure` for async functions. Furthermore, what would be the value of `isRefreshed` if `getCause` is not `null`? We can come up with some rules about which flag is true/false/null in relation to others, but I doubt that this would be an improvement over checking the `refreshTokenDocumentId` for `null` that this thread is arguing against.

@jkakavas 
>why don't we like all integrity and null checks in one place?

I don't like to have any checks far from the code relying on those checks, because when code changes in one place it has to change in the other, and I could more easily forget to change it in two places rather than one. Moreover, it is not clear to me what is the _span_ that a "previous" check covers. The guarantees of a previous check _fade away_ as an unmutable object is passed around in functions that could change it. For example, we make some guarantees on the document write function a few lines above, and then read the doc here, and do the checks. Why bother with the checks, then, if they were asserted "before", in the write method?

> I find it preferable, compared to checking before use in many places

I disagree. In a world with `null` "values" and no immutable objects, I would be checking for `null` on every single step. **This world is real**, though, it's the world in which all objects are `Map`s of `Map`s . Fortunately we can break free of that world by parsing the map in an entity object, that has **no** `null` "values". **This world is indeed not real**, because the Java gods have blessed us with the `null` "value", so we decided to use `null`s in very "specific" circumstances, where "specific" is not defined (can be `null`).